### PR TITLE
AssetsDefinition.map_asset_specs

### PR DESCRIPTION
**Pull request by sryza**
## Summary & Motivation

Adds a utility that makes it easier to replace asset-level attributes on `AssetsDefinition`s.

### CODE REVIEW

File: python_modules/dagster/dagster/_core/definitions/assets.py [hunk 1]
```diff
@@ -1304,6 +1305,21 @@ def update_replace_dict_and_conflicts(
...
         return self.__class__.dagster_internal_init(**merged_attrs)
 
+    def map_asset_specs(self, fn: Callable[[AssetSpec], AssetSpec]) -> "AssetsDefinition":
+        mapped_specs = []
+        for spec in self.specs:
+            mapped_spec = fn(spec)
+            if mapped_spec.key != spec.key:
+                raise DagsterInvalidDefinitionError(
+                    f"Asset key {spec.key.to_user_string()} was changed to "
+                    f"{mapped_spec.key.to_user_string()}. Mapping function must not change keys."
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/assets.py:1315]: should we maybe prevent it from changing deps as well for now? can always relax that in the future

**Review by OwenKephart — APPROVED** [review]
The sun shines a bit brighter this day

File: python_modules/dagster/dagster/_core/definitions/assets.py (hunk 1 above)
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/assets.py:1315]: great point

File: python_modules/dagster/dagster/_core/definitions/assets.py [hunk 2]
```diff
@@ -1304,6 +1305,30 @@ def update_replace_dict_and_conflicts(
...
+                )
+            if (
+                # check reference equality first for performance
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/assets.py:1318]: ran a little benchmark to see if short circuiting list equality for reference equality actually helps:
```python
from datetime import datetime
from random import random


def random_list(n):
    return [random() for _ in range(n)]


random_lists = [random_list(1000) for _ in range(10000)]

start = datetime.now()
for _ in range(1000):
    for lst in random_lists:
        assert lst is lst or lst == lst


end = datetime.now()
print(end.timestamp() - start.timestamp())
```

short-circuiting with reference equality checks resulted in 10x improvment

**Pull request merged**